runtime/vm/ContinuationHelpers.cpp

```diff
@@ -90,28 +90,46 @@ createContinuation(J9VMThread *currentThread, j9object_t continuationObject)
+j9object_t
+synchronizeWithConcurrentGCScan(J9VMThread *currentThread, j9object_t continuationObject, J9VMContinuation *continuation)
 {
-	volatile uintptr_t *localAddr = &continuation->state;
+
 	/* atomically 'or' (not 'set') continuation->state  with currentThread */
-	uintptr_t oldContinuationState = VM_AtomicSupport::bitOr(localAddr, (uintptr_t)currentThread);
+	volatile uintptr_t *localAddr = &continuation->state;
 
-	Assert_VM_Null(VM_VMHelpers::getCarrierThreadFromContinuationState(oldContinuationState));
+	uintptr_t oldContinuationState = *localAddr;
+	uintptr_t newContinuationState = VM_VMHelpers::setCarrierThreadToContinuationState(oldContinuationState, currentThread) | VM_VMHelpers::getPendingToBeMountedMask();
+	uintptr_t returnContinuationState = oldContinuationState;
+	while ((returnContinuationState = VM_AtomicSupport::lockCompareExchange(localAddr, oldContinuationState, newContinuationState)) != oldContinuationState) {
+		oldContinuationState = *localAddr;
+		newContinuationState = VM_VMHelpers::setCarrierThreadToContinuationState(oldContinuationState, currentThread) | VM_VMHelpers::getPendingToBeMountedMask();
+	}
 
-	if (VM_VMHelpers::isConcurrentlyScannedFromContinuationState(oldContinuationState)) {
-		/* currentThread was low tagged (GC was already in progress), but by 'or'-ing our ID, we let GC know there is a pending mount */
-		internalReleaseVMAccess(currentThread);
+	Assert_VM_Null(VM_VMHelpers::getCarrierThreadFromContinuationState(returnContinuationState));
 
-		omrthread_monitor_enter(currentThread->publicFlagsMutex);
-		while (VM_VMHelpers::isConcurrentlyScannedFromContinuationState(*localAddr)) {
-			/* GC is still concurrently scanning the continuation(currentThread was still low tagged), wait for GC thread to notify us when it's done. */
-			omrthread_monitor_wait(currentThread->publicFlagsMutex);
+	do {
+		if (VM_VMHelpers::isConcurrentlyScannedFromContinuationState(returnContinuationState)) {
+			/* currentThread was low tagged (GC was already in progress), but by 'or'-ing our ID, we let GC know there is a pending mount */
+
+			PUSH_OBJECT_IN_SPECIAL_FRAME(currentThread, continuationObject);
+			internalReleaseVMAccess(currentThread);
+
+			omrthread_monitor_enter(currentThread->publicFlagsMutex);
+			while (VM_VMHelpers::isConcurrentlyScannedFromContinuationState(*localAddr)) {
+				/* GC is still concurrently scanning the continuation(currentThread was still low tagged), wait for GC thread to notify us when it's done. */
+				omrthread_monitor_wait(currentThread->publicFlagsMutex);
+			}
+			omrthread_monitor_exit(currentThread->publicFlagsMutex);
+
+			internalAcquireVMAccess(currentThread);
+			continuationObject = POP_OBJECT_IN_SPECIAL_FRAME(currentThread);
 		}
-		omrthread_monitor_exit(currentThread->publicFlagsMutex);
+		oldContinuationState = *localAddr;
+		newContinuationState = oldContinuationState & ~VM_VMHelpers::getPendingToBeMountedMask();
```

Comment:
to hide details how we maintain pending bit, create a helper resetPendingToContinuationState